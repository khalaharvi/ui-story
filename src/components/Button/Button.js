import React from 'react';
import PropTypes from 'prop-types';
import styled from 'react-emotion';
import _get from 'lodash.get';

const styles = ({
  primary,
  secondary,
  inverted,
  disabled,
  icon,
  iconAfter,
  white,
  theme
}) => {
  let backgroundColor;
  let color;
  let hoverStyles;
  let borderStyles;

  if (!disabled) {
    if (primary) {
      hoverStyles = {
        backgroundColor: _get(theme, 'buttons.primary.hover.backgroundColor'),
        borderColor: _get(theme, 'buttons.primary.hover.borderColor'),
        color: _get(theme, 'buttons.primary.hover.color')
      };

      if (inverted) {
        hoverStyles = {
          backgroundColor: _get(theme, 'buttons.primaryInverted.hover.backgroundColor'),
          color: _get(theme, 'buttons.primaryInverted.hover.color')
        };
      }

      if (white) {
        hoverStyles = {
          backgroundColor: _get(theme, 'buttons.primaryWhite.hover.backgroundColor'),
          color: _get(theme, 'buttons.primaryWhite.hover.color')
        };
      }
    }

    if (secondary) {
      hoverStyles = {
        borderColor: _get(theme, 'buttons.secondary.hover.borderColor')
      };
    }
  }

  if (primary) {
    backgroundColor = _get(theme, 'buttons.primary.backgroundColor');
    color = _get(theme, 'buttons.primary.color');
    borderStyles = {
      border: _get(theme, 'buttons.primary.border')
    };

    if (disabled) {
      backgroundColor = _get(theme, 'buttons.primary.disabled.backgroundColor');
      borderStyles = {};
      color = _get(theme, 'buttons.primary.disabled.color');
    }

    if (!disabled && inverted) {
      backgroundColor = _get(theme, 'buttons.primaryInverted.backgroundColor');
      borderStyles = {
        border: _get(theme, 'buttons.primaryInverted.border')
      };
      color = _get(theme, 'buttons.primaryInverted.color');
    }

    if (!disabled && white) {
      backgroundColor = _get(theme, 'buttons.primaryWhite.backgroundColor');
      borderStyles = {
        border: _get(theme, 'buttons.primaryWhite.border')
      };
      color = _get(theme, 'buttons.primaryWhite.color');
    }
  }

  if (secondary) {
    backgroundColor = _get(theme, 'buttons.secondary.backgroundColor');
    color = _get(theme, 'buttons.secondary.color');

    borderStyles = {
      border: `1px solid ${_get(theme, 'buttons.secondary.borderColor')}`,
      borderColor: _get(theme, 'buttons.secondary.borderColor')
    };

    if (disabled) {
      color = _get(theme, 'buttons.secondary.disabled.color');

      borderStyles = {
        ...borderStyles,
        borderColor: _get(theme, 'buttons.secondary.disabled.borderColor')
      };
    }
  }

  return {
    backgroundColor,
    color,
    lineHeight: disabled && !secondary ? '45px' : '43px',
    cursor: disabled ? 'not-allowed' : 'pointer',
    '&:hover': hoverStyles,
    ...borderStyles,

    display: icon ? 'flex' : undefined,
    alignItems: icon ? 'center' : undefined,
    svg: {
      margin: iconAfter ? '0 0 0 5px' : '0 5px 0 0',
      fill: 'currentColor'
    }
  };
};

const Button = styled('button')({
  padding: '0 20px',
  appearance: 'none',
  border: 0,
  outline: 0,
  background: 'none',
  cursor: 'pointer',
  lineHeight: '45px'
}, styles);

const ButtonComponent = ({
  disabled,
  loading,
  children,
  icon: Icon,
  iconAfter,
  ...rest
}) => (
  <Button disabled={disabled || loading} icon={!!Icon} iconAfter={iconAfter} {...rest}>
    { Icon && !iconAfter ? Icon : null }
    <span>{children}</span>
    { Icon && iconAfter ? Icon : null }
  </Button>
);

ButtonComponent.propTypes = {
  disabled: PropTypes.bool,
  hidden: PropTypes.bool,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
  primary: PropTypes.bool,
  white: PropTypes.bool,
  secondary: PropTypes.bool,
  inverted: PropTypes.bool,
  icon: PropTypes.element,
  iconAfter: PropTypes.bool
};

ButtonComponent.defaultProps = {
  disabled: false,
  hidden: false,
  loading: false,

  inverted: false,
  white: false,
  primary: false,
  secondary: false,
  iconAfter: false
};

export default ButtonComponent;
