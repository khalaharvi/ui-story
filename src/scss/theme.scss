@import 'variables';

// https://github.com/sass-mq/sass-mq
@import 'sass-mq/mq';

@import 'placeholders';
@import 'mixins';
@import 'functions';
@import 'columns';

// Templates
@import 'templates/commerce-module';
@import 'templates/facets';
@import 'templates/header';
@import 'templates/product-detail';
@import 'templates/products-results';
@import 'templates/products-search';
@import 'templates/related-module';
@import 'templates/search-results-header';
@import 'templates/seller-profile';
@import 'templates/sellers-results';
@import 'templates/sellers-search';

// Components
@import 'components/active-facets';
@import 'components/breadcrumb';
@import 'components/button';
@import 'components/clear-facets';
@import 'components/commerce-module';
@import 'components/conversion-buttons';
@import 'components/error';
@import 'components/facets-toggler';
@import 'components/geosearch';
@import 'components/icons';
@import 'components/links';
@import 'components/list-facet';
@import 'components/nav-bar';
@import 'components/panel';
// @import 'components/place-autocomplete';
@import 'components/powered-by';
@import 'components/product-detail';
@import 'components/products-hits';
@import 'components/range-facet';
@import 'components/search-autocomplete';
@import 'components/search-box';
@import 'components/section-header';
@import 'components/section-nav';
@import 'components/seller-profile';
@import 'components/seller-profile-memberships';
@import 'components/seller-profile-news';
@import 'components/seller-profile-portfolios';
@import 'components/sellers-hits';
@import 'components/sortby-facet';
@import 'components/toggler-facet';

// Block page scroll when facets are visible
body.#{$namespace}-facets-visible {
  overflow: hidden;
  height: 100%;
  width: 100%;
}

[hidden] {
  display: none !important;
}

input[type="search"]::-webkit-search-decoration,
input[type="search"]::-webkit-search-cancel-button,
input[type="search"]::-webkit-search-results-button,
input[type="search"]::-webkit-search-results-decoration {
  display: none;
}

::placeholder {
  color: $color__placeholder;
}
