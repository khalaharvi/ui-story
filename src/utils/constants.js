export const PRIMARY_CATEGORY_KEY = 'categories.lvl0';
export const SECONDARY_CATEGORY_KEY = 'categories.lvl1';
export const TERTIARY_CATEGORY_KEY = 'categories.lvl2';
