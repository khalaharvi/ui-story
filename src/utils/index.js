export { default as createClassNames } from './createClassNames';
export { default as getDisplayName } from './getDisplayName';
export { default as capitalizeFirstLetter } from './capitalizeFirstLetter';
export { default as slugify } from './slugify';
