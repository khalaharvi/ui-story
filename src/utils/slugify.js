export default (text) => {
  // Use hash map for special characters
  const specialChars = {"à":'a',"ä":'a',"á":'a',"â":'a',"æ":'a',"å":'a',"ë":'e',"è":'e',"é":'e', "ê":'e',"î":'i',"ï":'i',"ì":'i',"í":'i',"ò":'o',"ó":'o',"ö":'o',"ô":'o',"ø":'o',"ù":'o',"ú":'u',"ü":'u',"û":'u',"ñ":'n',"ç":'c',"ß":'s',"ÿ":'y',"œ":'o',"ŕ":'r',"ś":'s',"ń":'n',"ṕ":'p',"ẃ":'w',"ǵ":'g',"ǹ":'n',"ḿ":'m',"ǘ":'u',"ẍ":'x',"ź":'z',"ḧ":'h',"·":'-',"/":'-',"_":'-',",":'-',":":'-',";":'-'}; // eslint-disable-line

  return text.toString().toLowerCase()
    .replace(/\s+/g, '-') // Replace spaces with -
    .replace(/./g, target => specialChars[target] || target) // Replace special characters using the hash map
    .replace(/&/g, '-and-') // Replace & with 'and'
    .replace(/[^\w-]+/g, '') // Remove all non-word chars
    .replace(/--+/g, '-') // Replace multiple - with single -
    .replace(/^-+/, '') // Trim - from start of text
    .replace(/-+$/, ''); // Trim - from end of text
};
