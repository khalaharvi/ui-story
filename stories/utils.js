/* eslint-disable import/prefer-default-export */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { InstantSearch } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';
import { withMedia, MediaQueryProvider } from 'react-media-query-hoc';
import { withTests as withTestsAddon } from '@storybook/addon-jest';
import { ThemeProvider } from 'emotion-theming';
import dhTheme from '@dering-hall/dh-theme';
import testsResults from '../.jest-test-results.json';

const Wrap = ({ index, children }) => (
  <InstantSearch
    appId={process.env.ALGOLIA_ID}
    apiKey={process.env.ALGOLIA_API_KEY}
    indexName={index}
  >
    {children}
  </InstantSearch>
);

Wrap.propTypes = {
  index: PropTypes.string,
  children: PropTypes.node
};

Wrap.defaultProps = {
  index: process.env.ALGOLIA_PRODUCT_INDEX
};

const WrapThemeProvider = ({ children, theme }) => (
  <ThemeProvider theme={theme}>
    <Fragment>
      {children}
    </Fragment>
  </ThemeProvider>
);

WrapThemeProvider.propTypes = {
  theme: PropTypes.object // eslint-disable-line
};

WrapThemeProvider.defaultProps = {
  theme: dhTheme
};

const SearchStateDebug = connectStateResults(({ searchState }) => (
  <div style={{ margin: '1rem 0', fontFamily: 'monospace' }}>
    <pre
      style={{
        background: '#f6f8fa',
        fontSize: '.65rem',
        padding: '.5rem'
      }}
    >
      <strong>searchState</strong> = {JSON.stringify(searchState, null, '\t')}
    </pre>
  </div>
));

const customQueries = {
  small: 'screen and (max-width: 730px)',
  medium: 'screen and (max-width: 970px)',
  large: 'screen and (max-width: 1240px)',
  xlarge: 'screen and (min-width: 1240px)'
};

const MediaQueriesWrapper = ({ children }) => (
  <MediaQueryProvider queries={customQueries}>
    {children}
  </MediaQueryProvider>
);

const OnSmallOnly = withMedia(({ media, children }) => {
  if (!media.small) {
    return null;
  }

  return (
    <Fragment>
      {children}
    </Fragment>
  );
});

const withTests = withTestsAddon({
  results: testsResults
});

export {
  SearchStateDebug,
  MediaQueriesWrapper,
  OnSmallOnly,
  Wrap,
  WrapThemeProvider,
  withTests
};
/* eslint-enable */
