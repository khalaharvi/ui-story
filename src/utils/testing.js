import { object } from 'prop-types';
import { channel, createBroadcast } from 'emotion-theming';
import { shallow, mount } from 'enzyme';
import theme from '@dering-hall/dh-theme';

const getOptions = (options = {}) => {
  const broadcast = createBroadcast(theme);

  return {
    context: {
      [channel]: broadcast
    },
    childContextTypes: {
      [channel]: object
    },
    ...options
  };
};

export const mountWithTheme = (
  Component,
  options = {}
) => mount(Component, { ...getOptions(options) });

export const shallowWithTheme = (
  Component,
  options = {}
) => shallow(Component, { ...getOptions(options) });

export default {
  mountWithTheme,
  shallowWithTheme
};
