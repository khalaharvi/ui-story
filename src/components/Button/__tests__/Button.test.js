import React from 'react';
import { mount } from 'enzyme';
import { mountWithTheme } from 'utils/testing';
import theme from '@dering-hall/dh-theme';

import Button from '../Button';

describe('<Button />', () => {
  it('renders correctly as primary', () => {
    const wrapper = mount(<Button theme={theme} primary>Primary Button</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly as primary and inverted', () => {
    const wrapper = mount(<Button theme={theme} primary inverted>Primary Inverted Button</Button>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const wrapper = mount(
      <Button theme={theme} primary className="my-btn">
        Primary Button with custom className
      </Button>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when disabled', () => {
    const wrapper = mount(
      <Button theme={theme} primary disabled>
        Disabled Primary Button
      </Button>
    );

    expect(
      wrapper
        .find('button')
        .prop('disabled')
    ).toBe(true);
    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('should be disabled when loading', () => {
    const wrapper = mount(
      <Button theme={theme} primary loading>Primary Button Loading...</Button>
    );

    expect(
      wrapper
        .find('button')
        .prop('disabled')
    ).toBe(true);

    wrapper.unmount();
  });

  it('renders correctly when hidden', () => {
    const wrapper = mount(
      <Button theme={theme} primary hidden>Primary Button Hidden</Button>
    );

    expect(
      wrapper
        .find('button')
        .prop('hidden')
    ).toBe(true);

    wrapper.unmount();
  });

  it('should call onClick when clicked', () => {
    const onClick = jest.fn();
    const wrapper = mount(
      <Button theme={theme} primary onClick={onClick}>Primary Button</Button>
    );

    const button = wrapper.find('button');

    button.simulate('click');

    expect(onClick.mock.calls).toHaveLength(1);

    wrapper.unmount();
  });

  it('can use theme from a ThemeProvider', () => {
    const wrapper = mountWithTheme(<Button primary>Themed using ThemeProvider</Button>);

    expect(wrapper).toMatchSnapshot();
  });
});
