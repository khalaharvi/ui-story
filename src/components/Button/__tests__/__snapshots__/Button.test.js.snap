// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> can use theme from a ThemeProvider 1`] = `
.emotion-0 {
  padding: 0 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: 0;
  background: none;
  cursor: pointer;
  line-height: 45px;
  background-color: transparent;
  color: #242424;
  line-height: 43px;
  cursor: pointer;
  border: 1px solid #242424;
}

.emotion-0:hover {
  background-color: #242424;
  border-color: #242424;
  color: #fff;
}

.emotion-0 svg {
  margin: 0 5px 0 0;
  fill: currentColor;
}

<ButtonComponent
  disabled={false}
  hidden={false}
  iconAfter={false}
  inverted={false}
  loading={false}
  primary={true}
  secondary={false}
  white={false}
>
  <Styled(button)
    disabled={false}
    hidden={false}
    icon={false}
    iconAfter={false}
    inverted={false}
    primary={true}
    secondary={false}
    white={false}
  >
    <button
      className="emotion-0 emotion-1"
      disabled={false}
      hidden={false}
    >
      <span>
        Themed using ThemeProvider
      </span>
    </button>
  </Styled(button)>
</ButtonComponent>
`;

exports[`<Button /> renders correctly as primary 1`] = `
.emotion-0 {
  padding: 0 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: 0;
  background: none;
  cursor: pointer;
  line-height: 45px;
  background-color: transparent;
  color: #242424;
  line-height: 43px;
  cursor: pointer;
  border: 1px solid #242424;
}

.emotion-0:hover {
  background-color: #242424;
  border-color: #242424;
  color: #fff;
}

.emotion-0 svg {
  margin: 0 5px 0 0;
  fill: currentColor;
}

<ButtonComponent
  disabled={false}
  hidden={false}
  iconAfter={false}
  inverted={false}
  loading={false}
  primary={true}
  secondary={false}
  white={false}
>
  <Styled(button)
    disabled={false}
    hidden={false}
    icon={false}
    iconAfter={false}
    inverted={false}
    primary={true}
    secondary={false}
    white={false}
  >
    <button
      className="emotion-0 emotion-1"
      disabled={false}
      hidden={false}
    >
      <span>
        Primary Button
      </span>
    </button>
  </Styled(button)>
</ButtonComponent>
`;

exports[`<Button /> renders correctly as primary and inverted 1`] = `
.emotion-0 {
  padding: 0 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: 0;
  background: none;
  cursor: pointer;
  line-height: 45px;
  background-color: #242424;
  color: #fff;
  line-height: 43px;
  cursor: pointer;
  border: 1px solid #242424;
}

.emotion-0:hover {
  background-color: transparent;
  color: #242424;
}

.emotion-0 svg {
  margin: 0 5px 0 0;
  fill: currentColor;
}

<ButtonComponent
  disabled={false}
  hidden={false}
  iconAfter={false}
  inverted={true}
  loading={false}
  primary={true}
  secondary={false}
  white={false}
>
  <Styled(button)
    disabled={false}
    hidden={false}
    icon={false}
    iconAfter={false}
    inverted={true}
    primary={true}
    secondary={false}
    white={false}
  >
    <button
      className="emotion-0 emotion-1"
      disabled={false}
      hidden={false}
    >
      <span>
        Primary Inverted Button
      </span>
    </button>
  </Styled(button)>
</ButtonComponent>
`;

exports[`<Button /> renders correctly when disabled 1`] = `
.emotion-0 {
  padding: 0 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: 0;
  background: none;
  cursor: pointer;
  line-height: 45px;
  background-color: rgba(36,36,36,0.5);
  color: #fff;
  line-height: 45px;
  cursor: not-allowed;
}

.emotion-0 svg {
  margin: 0 5px 0 0;
  fill: currentColor;
}

<ButtonComponent
  disabled={true}
  hidden={false}
  iconAfter={false}
  inverted={false}
  loading={false}
  primary={true}
  secondary={false}
  white={false}
>
  <Styled(button)
    disabled={true}
    hidden={false}
    icon={false}
    iconAfter={false}
    inverted={false}
    primary={true}
    secondary={false}
    white={false}
  >
    <button
      className="emotion-0 emotion-1"
      disabled={true}
      hidden={false}
    >
      <span>
        Disabled Primary Button
      </span>
    </button>
  </Styled(button)>
</ButtonComponent>
`;

exports[`<Button /> renders correctly with className 1`] = `
.emotion-0 {
  padding: 0 20px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: 0;
  outline: 0;
  background: none;
  cursor: pointer;
  line-height: 45px;
  background-color: transparent;
  color: #242424;
  line-height: 43px;
  cursor: pointer;
  border: 1px solid #242424;
}

.emotion-0:hover {
  background-color: #242424;
  border-color: #242424;
  color: #fff;
}

.emotion-0 svg {
  margin: 0 5px 0 0;
  fill: currentColor;
}

<ButtonComponent
  className="my-btn"
  disabled={false}
  hidden={false}
  iconAfter={false}
  inverted={false}
  loading={false}
  primary={true}
  secondary={false}
  white={false}
>
  <Styled(button)
    className="my-btn"
    disabled={false}
    hidden={false}
    icon={false}
    iconAfter={false}
    inverted={false}
    primary={true}
    secondary={false}
    white={false}
  >
    <button
      className="my-btn emotion-0 emotion-1"
      disabled={false}
      hidden={false}
    >
      <span>
        Primary Button with custom className
      </span>
    </button>
  </Styled(button)>
</ButtonComponent>
`;
